source/common/network/io_socket_handle_impl.cc

```diff
@@ -0,0 +1,67 @@
+#include "common/network/io_socket_handle_impl.h"
+
+#include <errno.h>
+
+#include <iostream>
+
+using Envoy::Api::SysCallIntResult;
+using Envoy::Api::SysCallSizeResult;
+
+namespace Envoy {
+namespace Network {
+
+IoError::IoErrorCode IoSocketError::errorCode() const {
+  switch (errno_) {
+  case EAGAIN:
+    RELEASE_ASSERT(false, "EAGAIN should use specific error ENVOY_ERROR_AGAIN");
+    return IoErrorCode::Again;
+  case ENOTSUP:
+    return IoErrorCode::NoSupport;
+  case EAFNOSUPPORT:
+    return IoErrorCode::AddressFamilyNoSupport;
+  case EINPROGRESS:
+    return IoErrorCode::InProgress;
+  case EPERM:
+    return IoErrorCode::Permission;
+  default:
+    return IoErrorCode::UnknownError;
+  }
+}
+
+std::string IoSocketError::errorDetails() const { return ::strerror(errno_); }
+
+IoSocketHandleImpl::~IoSocketHandleImpl() {
+  if (fd_ != -1) {
+    IoSocketHandleImpl::close();
+  }
+}
+
+// Deallocate memory only if the error is not ENVOY_ERROR_AGAIN.
+void deleteIoError(IoError* err) {
+  ASSERT(err != nullptr);
+  if (err != ENVOY_ERROR_AGAIN) {
+    delete err;
+  }
+}
+
+IoHandleCallUintResult IoSocketHandleImpl::close() {
+  ASSERT(fd_ != -1);
+  const int rc = ::close(fd_);
+  fd_ = -1;
+  IoErrorPtr err(nullptr, deleteIoError);
+  if (rc == -1) {
```

Comment:
Can this ever actually fail? It definitely can't return EAGAIN I don't think. Can we just RELEASE_ASSERT that is doesn't fail instead of having this code?

Reply:
Yes, it couldn't be EAGAIN for sure. But it can fail according t linux man.

Reply:
I think it can only fail if passed a bad handle though, which should never happen. I would just RELEASE_ASSERT it personally.

Reply:
RELEASE_ASSERT will fail test/common/network/connection_impl_test.cc because most of the io_handle objects in the test are initialized manually with invalid fd like 0 or 1.
I removed the `if (rc == -1 )` block and return `rc` as is.